src/marker.rs

```diff
@@ -528,21 +286,133 @@ impl<'py> Python<'py> {
+
+        use crate::impl_::panic::PanicTrap;
+
         // Use a guard pattern to handle reacquiring the GIL,
         // so that the GIL will be reacquired even if `f` panics.
         // The `Send` bound on the closure prevents the user from
         // transferring the `Python` token into the closure.
         let _guard = unsafe { SuspendGIL::new() };
-        f()
+
+        // To close soundness loopholes w.r.t. `send_wrapper` or `scoped-tls`,
+        // we run the closure on a separate thread so that it cannot
+        // access thread-local storage from the current thread.
+
+        // 1. Construct a task
+        struct Task(*mut dyn FnMut());
+        unsafe impl Send for Task {}
+
+        let (result_sender, result_receiver) = sync_channel::<Result<T>>(0);
+
+        let mut f = Some(f);
+
+        let mut task = || {
+            let f = f.take().unwrap();
+
+            let result = catch_unwind(AssertUnwindSafe(f));
+
+            result_sender.send(result).unwrap();
+        };
+
+        // SAFETY: the current thread will block until the closure has returned
+        let task = Task(unsafe { transmute(&mut task as &mut dyn FnMut()) });
+
+        // 2. Dispatch task to waiting thread, spawn new thread if necessary
+        let trap = PanicTrap::new(
+            "allow_threads panicked while stack data was accessed by another thread which is a bug",
+        );
+
+        static THREADS: Mutex<Vec<SyncSender<Task>>> = const_mutex(Vec::new());
+
+        enum State {
+            Pending(Task),
+            Dispatched(SyncSender<Task>),
+        }
+
+        let mut state = State::Pending(task);
+
+        while let Some(task_sender) = THREADS.lock().pop() {
+            match state {
+                State::Pending(task) => match task_sender.send(task) {
+                    Ok(()) => {
+                        state = State::Dispatched(task_sender);
+                        break;
+                    }
+                    Err(SendError(task)) => {
+                        state = State::Pending(task);
+                        continue;
+                    }
+                },
+                State::Dispatched(_task_sender) => unreachable!(),
+            }
+        }
+
+        let task_sender = match state {
+            State::Pending(task) => {
+                let (task_sender, task_receiver) = sync_channel::<Task>(0);
+
+                spawn(move || {
+                    while let Ok(task) = task_receiver.recv_timeout(Duration::from_secs(60)) {
+                        // SAFETY: all data accessed by `task` will stay alive until it completes
+                        unsafe { (*task.0)() };
+                    }
+                });
+
+                task_sender.send(task).unwrap();
+
+                task_sender
+            }
+            State::Dispatched(task_sender) => task_sender,
+        };
+
+        // 3. Wait for completion and check result
+        let result = result_receiver.recv().unwrap();
```

Comment:
Yes, if the worker thread died and disconnected the channel without a value which should not happen because we catch panics from the closure.

And indeed, this will trigger the panic trap and abort with the corresponding error message.

Generally, I went for more rather than fewer `unwrap`s to increase the chances that we actually hit our panic trap if anything unexpected happens rather than silently take the error and hit some unexpected implied condition. Mainly the because some downstream inconsistencies often leads to debugging sessions beginning with a question "But this state should not be possible?!".

Reply:
Agreed that this should panic, the only question here imo is whether this should be `.panic()` or `.expect("worker thread died unexpectedly")`

Reply:
I think `.expect` is basically always better than `.unwrap` and it is just laziness to not do that. The only downside I see is that the code could become cluttered eventually, but since this part is so delicate, I think the additional documentation and debuggability are definitely worth at little clutter. Pushed a version which documents all `unwrap`s.